library/std/src/ffi/os_str.rs

```diff
@@ -1101,6 +1101,40 @@ impl<'a> From<Cow<'a, OsStr>> for OsString {
     }
 }
 
+/// An error returned by `[TryFrom::<&OsStr>::try_from]` for [`&str`].
+#[stable(feature = "str_tryfrom_osstr_impl", since = "1.64.0")]
+#[derive(Debug)]
+#[non_exhaustive]
+pub struct StrFromOsStrError {}
```

Comment:
In terms of error information density, I wonder if we shouldn't set the bar higher. There is a lot more information in a [`UTF8Error`](https://doc.rust-lang.org/stable/std/str/struct.Utf8Error.html), and we already create and discard one on at least [some platforms](https://github.com/rust-lang/rust/blob/master/library/std/src/sys/unix/os_str.rs#L199) with the current implementation. Maybe we should skip relying on `OsStr::to_str` and build something more fundamental here.

Reply:
I originally considered that as well, but I think might make too many assumptions about the internal representation of `OsStr`. Some future platform might choose a non-UTF-8 like encoding. On such a platform it may not make much sense to talk about specific byte locations of errors. It might not even on current platforms since (as far as I'm aware) there's no way to look at the byte representation of an `OsStr`. Therefore I chose the more conservative approach at first.
However if the consensus is to implement it with `Utf8Error` as the error type, I'd be open to implementing that as well.

Reply:
I think it should still be an opaque struct like you've done here, and we should probably document that the internal representation is going to be inherently unstable and is kept intentionally opaque but we can still have that internally be a `UTF8Error` when we're on unix and give nice utf8 related error messages, or we could do some windows specific wtf8 error about it containing surrogate pairs or w/e level of context we want to expose there, and so on.

Reply:
I think we're pretty constrained in what `OsStr` can be. The [`impl AsRef<OsStr> for str`](https://doc.rust-lang.org/std/primitive.str.html#impl-AsRef%3COsStr%3E) implies that every valid `str` must also be (bit-for-bit) a valid `OsStr`. So if nothing else, surely this `TryFrom` can report the valid UTF-8 span and the length of the invalid span, without providing details of why a particular sequence of bytes is invalid?

Reply:
Good point. This basically means that an `OsStr` must always have something like `[u8]` as the underlying representation and therefore it talking about sub-spans of the string should always make sense.
So should the error type simply be `Utf8Error` or some opaque error wrapping a `Utf8Error` and exposing some of it's details?

Reply:
My instinct is still to prefer an opaque independent struct with exposing some of it's details, since this should be more flexible going forward, but I am not sure if I'm being overly cautious here.

Reply:
Hm, did the Error API not get considered as part of the FCP?

Reply:
I think we could extend this as further work. It seems very reasonable to either expose the utf8 error, or something equivalent.

That said, prior to it being available as bytes, doing so may leak details about our internal representation, (on windows platforms for example I believe this would be the case) so *personally* I think we should wait until https://github.com/rust-lang/rust/pull/95290 is more settled.

This also avoids needing to do further API team work on the design of the error type's API, which *I think* means it could just land as is.

Reply:
(To be clear, I'm not suggesting that blocks landing this, just that it would block exposing information about in which way the UTF 8 bytes are invalid)

Reply:
@thomcc No the error type was not part of the FCP. In retrospective it would have been useful to make that part of it. Do I understand your suggestion correctly that you want to keep the `UTF8Error` type as the error type but with the methods on that type returning dummy values instead, so that users cannot rely on the internal representation? In that case that should probably be documented.

Reply:
No, I don't think that would be good. I think exposing an opaque struct with no methods aside from trait impls would be better. I'm not t-libs-api though.